File: planet-id-demo-rp-ios/PlanetIDRP/PlanetIDRP/PlanetID.swift
//
//  PlanetID.swift
//  PlanetIDRP
//
//  Created by Toomas Laigna on 13.02.2020.
//  Copyright © 2020 Planetway. All rights reserved.
//

open class PlanetID {
    public static func getAuthResponse(_ urlComponents: URLComponents) -> AuthResponse? {
        guard let query = urlComponents.queryItems else { return nil }
        
        guard let code = (query.first { (item) -> Bool in item.name == "code" }?.value) else { return nil }
        let state = query.first { (item) -> Bool in item.name == "state" }?.value
        guard let callback = (query.first { (item) -> Bool in item.name == "callback" }?.value) else { return nil }
        
        return AuthResponse(
            code: code,
            state: state,
            callback: callback
        )
    }
    
    public static func getError(_ urlComponents: URLComponents) -> String? {
        guard let query = urlComponents.queryItems else { return nil }
        let error = query.first { (item) -> Bool in item.name == "error" }
        return error?.value
    }

    public static func isRejected(_ urlComponents: URLComponents) -> Bool {
        guard let query = urlComponents.queryItems else { return false }
        let rejected = query.first { (item) -> Bool in item.name == "rejected" }
        return rejected != nil
    }
}

File: planet-id-demo-rp-android/app/src/main/java/com/planetway/planetid/rpsdk/PlanetId.kt
package com.planetway.planetid.rpsdk

import android.content.Context
import android.content.Intent
import android.net.Uri

object PlanetId {

    fun invokeAction(context: Context, authRequest: AuthRequest, returnAppUri: String) {
        val uriBuilder = Uri.Builder()
        uriBuilder.scheme("planetid")
        uriBuilder.authority("invoke-action")

        uriBuilder.appendQueryParameter("client_id", authRequest.clientId)
        uriBuilder.appendQueryParameter("signature_type", authRequest.action)
        uriBuilder.appendQueryParameter("scope", authRequest.scope)
        uriBuilder.appendQueryParameter("action", authRequest.action)
        uriBuilder.appendQueryParameter("redirect_uri", authRequest.redirectUri)
        uriBuilder.appendQueryParameter("response_type", authRequest.responseType)
        uriBuilder.appendQueryParameter("nonce", authRequest.nonce)
        uriBuilder.appendQueryParameter("state", authRequest.state)
        uriBuilder.appendQueryParameter("language", authRequest.language)
        uriBuilder.appendQueryParameter("return_app_uri", returnAppUri)
        if (authRequest.payload != null) {
            uriBuilder.appendQueryParameter("payload", authRequest.payload)
        }

        context.startActivity(
            Intent(Intent.ACTION_VIEW, uriBuilder.build()).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
            }
        )
    }

    fun getAuthResponse(uri: Uri): AuthResponse {
        val code = uri.getQueryParameter("code")!!
        val state = uri.getQueryParameter("state")
        val callback = uri.getQueryParameter("callback")!!
        return AuthResponse(
            code,
            state,
            callback
        )
    }

    fun isRejected(uri: Uri): Boolean {
        return uri.getQueryParameter("rejected") != null
    }

    fun getError(uri: Uri): String? {
        return uri.getQueryParameter("error")
    }
}
